PayloadPkg/FirmwareUpdate/FirmwareUpdateHelper.c

```diff
@@ -1033,6 +1036,66 @@ CheckAcmSvn (
   return EFI_SUCCESS;
 }
 
+  /**
+    Get cpu ucode revision.
+
+    @retval   cpu ucode revision
+  **/
+  UINT32
+  EFIAPI
+  GetCpuUCodeRev (
+    VOID
+  )
+  {
+    UINT64 MsrValue;
+    UINT32 UcodeRev;
+
+    AsmWriteMsr64 (MSR_IA32_BIOS_SIGN_ID, 0LL);
+    AsmCpuid (0, NULL, NULL, NULL, NULL);
+    MsrValue = AsmReadMsr64 (MSR_IA32_BIOS_SIGN_ID);
+    UcodeRev = RShiftU64 (MsrValue, 32) & 0xffffffff;
+
+    return UcodeRev;
+  }
+
+  /**
+    Perform UCODE revision check
+
+    This function will perform revision checks for UCODE in flash and
+    UCODE in capsule.
+
+    @param[in]  ImageHdr       Pointer to fw mgmt capsule Image header
+
+    @retval  EFI_SUCCESS      revision check successful.
+    @retval  other            error occurred during firmware update
+  **/
+  EFI_STATUS
+  EFIAPI
+  CheckUCodeVersion (
+    IN   EFI_FW_MGMT_CAP_IMAGE_HEADER  *ImageHdr
+    )
+  {
+    UINT32                ExistingUCodeRev;
+    UINT32                NewUCodeRev;
+    CPU_MICROCODE_HEADER  *NewUCodeHdr;
+
+    if ((ImageHdr == NULL)) {
+      return EFI_INVALID_PARAMETER;
+    }
+
+    ExistingUCodeRev = GetCpuUCodeRev();
+
+    NewUCodeHdr = (CPU_MICROCODE_HEADER *)((UINTN)ImageHdr + sizeof(EFI_FW_MGMT_CAP_IMAGE_HEADER));
+    NewUCodeRev = NewUCodeHdr->UpdateRevision;
```

Comment:
In the SBL region it might have multiple uCode patches exist. Here it only checks the version of the first uCode patch.
To support multiple uCode case, maybe just get the biggest uCode version number from the capsule image as below?
UINT32
GetUcodeVer (
  IN  EFI_FW_MGMT_CAP_IMAGE_HEADER  *ImageHdr
  )
{
  UINTN                 ImageBase;
  UINTN                 Offset;
  CPU_MICROCODE_HEADER  *UCodeHdr;
  UINT8                 *ImageByte;
  UINT32                uCodeVer;

  if (PcdGet32 (PcdUcodeSlotSize) == 0) {
    return 0;
  }

  uCodeVer  = 0;
  Offset    = 0;
  ImageBase = (UINTN)ImageHdr + sizeof(EFI_FW_MGMT_CAP_IMAGE_HEADER);
  ImageByte = (UINT8*)(ImageBase + Offset);
  while (*ImageByte != PAD_BYTE && Offset < ImageHdr->UpdateImageSize) {
    UCodeHdr = (CPU_MICROCODE_HEADER *)ImageByte;

    // Ensure uCode size from header does not exceed slot size
    if (UCodeHdr->TotalSize > PcdGet32 (PcdUcodeSlotSize)) {
      uCodeVer = 0;
	  break;
    }

    if (UCodeHdr->UpdateRevision > uCodeVer) {
      uCodeVer = UCodeHdr->UpdateRevision;
    }

    Offset   += PcdGet32(PcdUcodeSlotSize);
    ImageByte = (UINT8*)(ImageBase + Offset);
  }

  return uCodeVer;
}